includes/api/class-wc-admin-rest-reports-stock-controller.php

```diff
@@ -0,0 +1,354 @@
+					'compare' => 'BETWEEN',
+					'type'    => 'NUMERIC',
+				),
+				array(
+					'key'   => '_stock_status',
+					'value' => 'instock',
+				),
+			);
+		} elseif ( in_array( $request['type'], array_keys( wc_get_product_stock_status_options() ), true ) ) {
+			$args['meta_query'] = array( // WPCS: slow query ok.
+				array(
+					'key'   => '_stock_status',
+					'value' => $request['type'],
+				),
+			);
+		}
+
+		return $args;
+	}
+
+	/**
+	 * Query products.
+	 *
+	 * @param  array $query_args Query args.
+	 * @return array
+	 */
+	protected function get_products( $query_args ) {
+		$query  = new WP_Query();
+		$result = $query->query( $query_args );
+
+		$total_posts = $query->found_posts;
+		if ( $total_posts < 1 ) {
+			// Out-of-bounds, run the query again without LIMIT for total count.
+			unset( $query_args['paged'] );
+			$count_query = new WP_Query();
+			$count_query->query( $query_args );
+			$total_posts = $count_query->found_posts;
+		}
+
+		return array(
+			'objects' => array_map( 'wc_get_product', $result ),
+			'total'   => (int) $total_posts,
+			'pages'   => (int) ceil( $total_posts / (int) $query->query_vars['posts_per_page'] ),
+		);
+	}
+
+	/**
+	 * Get all reports.
+	 *
+	 * @param  WP_REST_Request $request Request data.
+	 * @return array|WP_Error
+	 */
+	public function get_items( $request ) {
+		$query_args    = $this->prepare_reports_query( $request );
+		$query_results = $this->get_products( $query_args );
+
+		$objects = array();
+		foreach ( $query_results['objects'] as $object ) {
+			$data      = $this->prepare_item_for_response( $object, $request );
+			$objects[] = $this->prepare_response_for_collection( $data );
+		}
+
+		$page      = (int) $query_args['paged'];
+		$max_pages = $query_results['pages'];
+
+		$response = rest_ensure_response( $objects );
+		$response->header( 'X-WP-Total', $query_results['total'] );
+		$response->header( 'X-WP-TotalPages', (int) $max_pages );
+
+		$base = add_query_arg( $request->get_query_params(), rest_url( sprintf( '/%s/%s', $this->namespace, $this->rest_base ) ) );
+
+		if ( $page > 1 ) {
+			$prev_page = $page - 1;
+			if ( $prev_page > $max_pages ) {
+				$prev_page = $max_pages;
+			}
+			$prev_link = add_query_arg( 'page', $prev_page, $base );
+			$response->link_header( 'prev', $prev_link );
+		}
+		if ( $max_pages > $page ) {
+			$next_page = $page + 1;
+			$next_link = add_query_arg( 'page', $next_page, $base );
+			$response->link_header( 'next', $next_link );
+		}
+
+		return $response;
+	}
+
+	/**
+	 * Prepare a report object for serialization.
+	 *
+	 * @param  WC_Product      $product  Report data.
+	 * @param  WP_REST_Request $request Request object.
+	 * @return WP_REST_Response
+	 */
+	public function prepare_item_for_response( $product, $request ) {
+		$data = array(
+			'id'             => $product->get_id(),
+			'parent_id'      => $product->get_parent_id(),
+			'sku'            => $product->get_sku(),
```

Comment:
Since we are returning `sku` in the response, would it be okay to return the product name in the default object too? That would cut down on an additional query for us. I know this is kind of a debated topic but figured since we are already adding in the sku here, seems like the product name wouldn't be much more.

Reply:
I see no problem to include more fields here, for now I just put those in the design post. I'll include product's name.

Reply:
@Aljullu For variations will already include the attributes like in the design post.

Sample result using the WooCommerce CSV (I just updated a few quantities to test):

```json
[
  {
    "id": 34,
    "parent_id": 11,
    "name": "Hoodie - Blue, Yes",
    "sku": "woo-hoodie-blue-logo",
    "stock_status": "instock",
    "stock_quantity": 2,
    "_links": {
      "product": [
        {
          "href": "https:\/\/woo.local\/wp-json\/wc\/v3\/products\/11\/variations\/34"
        }
      ],
      "parent": [
        {
          "href": "https:\/\/woo.local\/wp-json\/wc\/v3\/products\/11"
        }
      ]
    }
  },
  {
    "id": 10,
    "parent_id": 0,
    "name": "V-Neck T-Shirt",
    "sku": "woo-vneck-tee",
    "stock_status": "instock",
    "stock_quantity": 2,
    "_links": {
      "product": [
        {
          "href": "https:\/\/woo.local\/wp-json\/wc\/v3\/products\/10"
        }
      ]
    }
  }
]
```

So those names should replace this IDs on your screenshot.

Note that `Hoodie - Blue, Yes` is a variable product `Hoodie - Blue, Yes`, since we already include the attributes, not sure if we need the parent column, also grouped products does not have parents anymore.

I noticed that there is a few links there, they are for editing? If is just for editing should be fine, it's possible to generate with the ID, but if is on frontend, we probably can include the product's permalink.

Or do you think that we should include more fields there.

Reply:
> Note that `Hoodie - Blue, Yes` is a variable product `Hoodie - Blue, Yes`, since we already include the attributes, not sure if we need the parent column, also grouped products does not have parents anymore.

@claudiosanches thanks for explaining that again! :slightly_smiling_face: 

There is still something I don't understand here. With the WooCommerce CSV products, it works as you described, but with products generated with `wc-smooth-generator`, it's not. I would say the difference is that the first ones use _[global attributes](https://docs.woocommerce.com/document/variable-product/#section-3)_ while the later ones use _custom attributes specific to product_.

As a prove of that: I have a product `perferendis unde atque qui qui` with variation `nesciunt`, `dolor` and `sed`:
![image](https://user-images.githubusercontent.com/3616980/49547695-0ca37300-f8a9-11e8-85e5-7d526659ae5d.png)
And this is what I get from the server:
![image](https://user-images.githubusercontent.com/3616980/49547735-2a70d800-f8a9-11e8-8830-06e666ada3ff.png)

As I said, with the product `Hoodie` it works as you described. The only difference I could see between them is that _Hoodie_ uses global attributes while `perferendis unde atque qui qui` uses a custom attribute.

> I noticed that there is a few links there, they are for editing? If is just for editing should be fine, it's possible to generate with the ID, but if is on frontend, we probably can include the product's permalink.
> 
> Or do you think that we should include more fields there.

That's a good point. I think we are not using any of those links neither in this report nor in any other, we usually generate them in the frontend. So I would say we can remove them, if nobody has anything against that.

Reply:
About custom attributes this is the way it works, on our current report system should display the same way, but now there we also display the SKU first.

> That's a good point. I think we are not using any of those links neither in this report nor in any other, we usually generate them in the frontend. So I would say we can remove them, if nobody has anything against that.

If you are talking about `_links`, this is just for the REST API, just providing a route if need to do more request, just a standard.

Reply:
> About custom attributes this is the way it works, on our current report system should display the same way, but now there we also display the SKU first.

That's unfortunate because we need some way to let the user distinguish between different variations. So is there no way to include the custom attributes in this endpoint? Otherwise we would need to make another query to populate the table.

> If you are talking about `_links`, this is just for the REST API, just providing a route if need to do more request, just a standard.

Sorry, I think I misunderstood your message before! The links we are displaying in the _Stock_ report all go to edit pages, so there is no need for a permalink.

Reply:
Based on my investigations for https://github.com/woocommerce/woocommerce/pull/22083, I found that if there are too many attributes (too many = more than 2) or attribute name has 2+ words, the variation name (aka post title) does not include any identification for the variation, just the parent's name.
 
Don't want to go directly into solution mode, but maybe if https://github.com/woocommerce/woocommerce/pull/22083 gets accepted and merged, we could use the string from the excerpt and expose it via API to identify the variation? (and it can be truncated on the front end if too long)

Reply:
This is a bigger conversation, something that we should not try to solve it here now, and maybe need to be patched on master.
Also including links to the product and SKU, is not complete required to list all attributes there, at least is what we have been doing on core.

Reply:
This information on variations title is something new, so maybe we can improve, probably Mike have better ideas about it too.